File: java/src/main/java/com/couchbase/grocery/models/GroceryItem.java
package com.couchbase.grocery.models;

import com.couchbase.lite.Document;
import com.couchbase.lite.MutableDocument;

/**
 * Inventory item. Field names match the seeded Capella documents 1:1 — see
 * the Android port's {@code GroceryItem.kt} which uses the same schema.
 */
public class GroceryItem {
    private String id;
    private String name;
    private String type;
    private double price;
    private String imageURL;      // capital URL — matches the doc key
    private int quantity;
    private Integer productId;
    private String sku;
    private String unit;

    public GroceryItem() {}

    public static GroceryItem fromDocument(Document doc) {
        GroceryItem item = new GroceryItem();
        item.id        = doc.getId();
        item.name      = doc.getString("name");
        item.type      = doc.getString("type");
        item.price     = doc.getDouble("price");
        item.imageURL  = doc.getString("imageURL");
        item.productId = doc.contains("productId") ? (int) doc.getLong("productId") : null;
        item.sku       = doc.getString("sku");
        item.unit      = doc.getString("unit");
        item.quantity  = readQuantity(doc);
        return item;
    }

    /**
     * Capella stores the live inventory count in {@code stockQty}. The seeded
     * documents and other platforms (web, iOS, Android, .NET) also keep a
     * separate {@code quantity} field as a P2P CRDT counter — we treat
     * {@code stockQty} as authoritative, then fall back to {@code quantity.value}
     * (CRDT dict shape) and finally to a flat {@code quantity} int for older
     * data.
     */
    private static int readQuantity(Document doc) {
        if (doc.contains("stockQty")) {
            int v = (int) doc.getLong("stockQty");
            if (v > 0) return v;
        }
        var crdt = doc.getDictionary("quantity");
        if (crdt != null && crdt.contains("value")) {
            return (int) crdt.getLong("value");
        }
        if (doc.contains("quantity")) {
            return (int) doc.getLong("quantity");
        }
        return 0;
    }

    public MutableDocument toMutable() {
        MutableDocument doc = id != null ? new MutableDocument(id) : new MutableDocument();
        doc.setString("name", name)
           .setString("type", type != null ? type : "")
           .setDouble("price", price)
           .setString("imageURL", imageURL != null ? imageURL : "")
           .setLong("stockQty", quantity);
        if (productId != null) doc.setLong("productId", productId);
        if (sku != null)       doc.setString("sku", sku);
        if (unit != null)      doc.setString("unit", unit);
        return doc;
    }

    public String getId()           { return id; }
    public void setId(String id)    { this.id = id; }
    public String getName()         { return name; }
    public String getType()         { return type; }
    public double getPrice()        { return price; }
    public String getImageURL()     { return imageURL; }
    public int getQuantity()        { return quantity; }
    public void setQuantity(int q)  { this.quantity = q; }
    public Integer getProductId()   { return productId; }
    public String getSku()          { return sku; }
    public String getUnit()         { return unit; }
}

File: Ionic/src/models/GroceryItem.ts
/**
 * Inventory item — field names match the seeded Capella documents (and the
 * Java / Android / iOS ports). Specifically:
 *   - `stockQty` is the authoritative quantity field. A separate `quantity`
 *     dict exists on some docs for P2P CRDT sync; we treat it as fallback.
 *   - `imageURL` uses capital URL (matches the doc key, not camelCase).
 */
export interface GroceryItem {
  id: string;
  name: string;
  type?: string;
  price: number;
  imageURL?: string;
  quantity: number;
  productId?: number;
  sku?: string;
  unit?: string;
}

/** Build a GroceryItem from a raw document body returned by CBL. */
export function groceryItemFromDoc(id: string, doc: Record<string, unknown>): GroceryItem {
  const quantity = readQuantity(doc);
  return {
    id,
    name: (doc.name as string) ?? '',
    type: doc.type as string | undefined,
    price: typeof doc.price === 'number' ? doc.price : 0,
    imageURL: doc.imageURL as string | undefined,
    quantity,
    productId: typeof doc.productId === 'number' ? doc.productId : undefined,
    sku: doc.sku as string | undefined,
    unit: doc.unit as string | undefined,
  };
}

function readQuantity(doc: Record<string, unknown>): number {
  const stockQty = doc.stockQty;
  if (typeof stockQty === 'number' && stockQty > 0) return stockQty;
  const crdt = doc.quantity as { value?: number } | undefined;
  if (crdt && typeof crdt.value === 'number') return crdt.value;
  if (typeof doc.quantity === 'number') return doc.quantity;
  return 0;
}
